### PR TITLE
[WIP] OpenIO C API ECD bypass

### PR DESCRIPTION
**Pull request by aneutron**
##### SUMMARY
The C API at the time of writing uses the ECD to download and upload content from EC-enabled containers.
The objective is then to do the encoding/decoding within the C client using _liberasurecode_.

##### COMPONENT NAME
`C API`

##### SDS VERSION
`openio 4.3.2.dev8`

##### ADDITIONAL INFORMATION
The download part of the pull request can be applied independently of the upload part.

The upload part requires `CURLOPT_TRAILERFUNC` and `CURLOPT_TRAILERDATA` options introduced in [this curl commit](https://github.com/curl/curl/commit/f464535bfdd9a83140d8a13c3fe3d937239d1c2a), and should be available in the `7.64.0` release.

### CODE REVIEW

File: core/sds.c [hunk 1]
```diff
@@ -806,6 +903,53 @@ _download_range_from_metachunk_ec(struct _download_ctx_s *dl,
...
+
+	gsize data_len;
+	const char *data = g_bytes_get_data(resulting_data, &data_len);
+
+	//FIXME: Not sure why there is a cast to unsigned char
+	int sent = dl->dst->data.hook.cb(dl->dst->data.hook.ctx,
+									 (const unsigned char*)data, data_len);
+
+	if (sent != (int)data_len)
+		err = NEWERROR(ERRCODE_UNKNOWN_ERROR, "Well, shit.");
```
Comment (fvennetier) [inline core/sds.c]: Again, this error message is not helpful.

File: core/sds.c [hunk 2]
```diff
@@ -966,7 +1110,10 @@ _download_to_hook (struct oio_sds_s *sds, struct oio_sds_dl_src_s *src,
 			.sds = sds, .dst = dst, .src = src, .chunk_method = chunk_method,
 			.metachunks = NULL, .chunks = chunks,
 		};
-		err = _organize_chunks(chunks, &dl.metachunks, sds->no_shuffle, 1);
+		int shuffle = _chunk_method_needs_ecd(dl.chunk_method) || sds->no_shuffle;
+		err = _organize_chunks(chunks, &dl.metachunks, shuffle, 1);
+		struct chunk_s *c = g_slist_nth_data(chunks, 0);
+		if(c){}
```
Comment (fvennetier) [inline core/sds.c]: These two lines seem of no use.

**Comment by jfsmig** [comment]
OB-87

**Comment by jfsmig** [comment]
PR rebased on master and moved as a feature branch named `ecd-bypass`

**Pull request closed without merging**